runtime/src/block_min_prioritization_fee_cache.rs

```diff
@@ -0,0 +1,595 @@
+use {
+    crate::{
+        block_min_prioritization_fee::*,
+        block_min_prioritization_fee_cache_query::BlockMinPrioritizationFeeCacheQuery,
+        block_min_prioritization_fee_cache_update::BlockMinPrioritizationFeeCacheUpdate,
+    },
+    solana_sdk::{
+        clock::Slot, pubkey::Pubkey, saturating_add_assign, transaction::SanitizedTransaction,
+    },
+    std::collections::HashMap,
+};
+
+/// The maximum number of blocks to keep in `BlockMinPrioritizationFeeCache`; States from
+/// up to 150 recent blocks should be sufficient to estimate minimal prioritization fee to
+/// land transactions to current block.
+const NUMBER_OF_RECENT_BLOCKS: usize = 150;
+
+/// Holds up to NUMBER_OF_RECENT_BLOCKS recent block's min prioritization fee for block,
+/// and for each writable accounts per block.
+pub struct BlockMinPrioritizationFeeCache {
+    cache: HashMap<Slot, BlockMinPrioritizationFee>,
+    // metrics counts
+    successful_transaction_update_count: u64,
+    fail_get_transaction_priority_details_count: u64,
+    fail_get_transaction_account_locks_count: u64,
+    fail_finalize_block_not_found: u64,
+}
```

Comment:
Would be good to add a timing field `total_update_elapsed_us: u64` and a metric for how long the multiple calls to `update_transactions` for a particular slot

Reply:
added it alone with other metrics refactoring